server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -249,74 +296,88 @@ public CachingClusteredClient(
 
-    Sequence<T> run(final UnaryOperator<TimelineLookup<String, ServerSelector>> timelineConverter)
+    Stream<Sequence<T>> run(final UnaryOperator<TimelineLookup<String, ServerSelector>> timelineConverter)
     {
       @Nullable
       TimelineLookup<String, ServerSelector> timeline = serverView.getTimeline(query.getDataSource());
       if (timeline == null) {
-        return Sequences.empty();
+        return Stream.empty();
       }
       timeline = timelineConverter.apply(timeline);
       if (uncoveredIntervalsLimit > 0) {
         computeUncoveredIntervals(timeline);
       }
 
-      final Set<ServerToSegment> segments = computeSegmentsToQuery(timeline);
+      Stream<ServerToSegment> segments = computeSegmentsToQuery(timeline);
       @Nullable
       final byte[] queryCacheKey = computeQueryCacheKey();
       if (query.getContext().get(QueryResource.HEADER_IF_NONE_MATCH) != null) {
+        // Materialize then re-stream
+        List<ServerToSegment> materializedSegments = segments.collect(Collectors.toList());
+        segments = materializedSegments.stream();
+
         @Nullable
         final String prevEtag = (String) query.getContext().get(QueryResource.HEADER_IF_NONE_MATCH);
         @Nullable
-        final String currentEtag = computeCurrentEtag(segments, queryCacheKey);
+        final String currentEtag = computeCurrentEtag(materializedSegments, queryCacheKey);
         if (currentEtag != null && currentEtag.equals(prevEtag)) {
-          return Sequences.empty();
+          return Stream.empty();
         }
       }
 
-      final List<Pair<Interval, byte[]>> alreadyCachedResults = pruneSegmentsWithCachedResults(queryCacheKey, segments);
-      final SortedMap<DruidServer, List<SegmentDescriptor>> segmentsByServer = groupSegmentsByServer(segments);
-      return new LazySequence<>(() -> {
-        List<Sequence<T>> sequencesByInterval = new ArrayList<>(alreadyCachedResults.size() + segmentsByServer.size());
-        addSequencesFromCache(sequencesByInterval, alreadyCachedResults);
-        addSequencesFromServer(sequencesByInterval, segmentsByServer);
-        return Sequences
-            .simple(sequencesByInterval)
-            .flatMerge(seq -> seq, query.getResultOrdering());
-      });
+      // This pipeline follows a few general steps:
+      // 1. Fetch cache results - Unfortunately this is an eager operation so that the non cached items can
+      // be batched per server. Cached results are assigned to a mock server ALREADY_CACHED_SERVER
+      // 2. Group the segment information by server
+      // 3. Per server (including the ALREADY_CACHED_SERVER) create the appropriate Sequence results - cached results
+      // are handled in their own merge
+      final Stream<SerializablePair<ServerToSegment, Optional<T>>> cacheResolvedResults = deserializeFromCache(
+          maybeFetchCacheResults(
+              queryCacheKey,
+              segments
+          )
+      );
+      return groupCachedResultsByServer(
```

Comment:
IMO, this style is far less readable than
```java
return groupCachedResultsByServer(cacheResolvedResults)
        .map(this::runOnServer)
        // Long comment here
        //
        //
        .collect(Collectors.toList())
        .stream()
```

Same throughout this PR

Reply:
I can do this in some place. In others, specifically with functions this kind of style becomes much harder to read. So just to keep things in the same format I used this style in all places

Reply:
Here's an example of the way lambdas screw with the current formatters:

```java
    private Stream<ServerToSegment> computeSegmentsToQuery(TimelineLookup<String, ServerSelector> timeline)
    {
      return toolChest
          .filterSegments(
              query,
              query.getIntervals().stream().flatMap(i -> timeline.lookup(i).stream()).collect(Collectors.toList())
          )
          .stream()
          .flatMap(
              holder -> DimFilterUtils.filterShards(
                  query.getFilter(),
                  holder.getObject(),
                  partitionChunk -> partitionChunk.getObject().getSegment().getShardSpec(),
                  Maps.newHashMap()
              )
                                      .stream()
                                      .map(
                                          chunk -> {
                                            ServerSelector server = chunk.getObject();
                                            final SegmentDescriptor segment = new SegmentDescriptor(
                                                holder.getInterval(),
                                                holder.getVersion(),
                                                chunk.getChunkNumber()
                                            );
                                            return new ServerToSegment(server, segment);
                                          }
                                      )
          )
          .distinct();
    }
```

Reply:
I'm trying to find a balance, will push a change with some targeted format changes

Reply:
The better formatting for you snippet is
```java
  private Stream<ServerToSegment> computeSegmentsToQuery(TimelineLookup<String, ServerSelector> timeline)
  {
    return toolChest
        .filterSegments(
            query,
            query.getIntervals().stream().flatMap(i -> timeline.lookup(i).stream()).collect(Collectors.toList())
        )
        .stream()
        .flatMap(holder -> DimFilterUtils
            .filterShards(
                query.getFilter(),
                holder.getObject(),
                partitionChunk -> partitionChunk.getObject().getSegment().getShardSpec(),
                Maps.newHashMap()
            )
            .stream()
            .map(chunk -> {
              ServerSelector server = chunk.getObject();
              final SegmentDescriptor segment = new SegmentDescriptor(
                  holder.getInterval(),
                  holder.getVersion(),
                  chunk.getChunkNumber()
              );
              return new ServerToSegment(server, segment);
            })
        )
        .distinct();
  }
```

However, note that "Effective Java" recommends to extract every lambda longer than 3 lines as a method. The above snippet contains two lambdas much longer than that: one of 16 lines and another of 7 lines. If they are extracted as methods,  Stream API code will look very accurate.

Another point: new IntelliJ [added type hints for Stream code](https://twitter.com/antonarhipov/status/1022139362736304129), that won't work in your formatting style, they only work in the "classical" formatting style.

Also, the classical style is predominately used in the Druid codebase already, e. g. in Query building code in tests.

Reply:
refactored some more to try and clean up formatting